### PR TITLE
Add branding-preference-mgt module and feature

### PR DESCRIPTION
**Pull request by UdeshAthukorala**
### Proposed changes in this pull request

Add branding preference-preference management module and feature.

### Follow up actions

- [ ] When merging, Update the versions in the pom files to the latest framework version.

### CODE REVIEW

File: components/branding-preference-mgt/org.wso2.carbon.identity.branding.preference.mgt/src/main/java/org/wso2/carbon/identity/branding/preference/mgt/internal/BrandingPreferenceManagerComponent.java [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+
+    private static final Log log = LogFactory.getLog(BrandingPreferenceManagerComponent.class);
+
+    @Activate
+    protected void activate(ComponentContext context) {
+
+        context.getBundleContext()
+                .registerService(BrandingPreferenceManager.class, new BrandingPreferenceManagerImpl(), null);
+
+        log.info("BrandingPreferenceMgtServiceComponent is activated.");
```
Comment (madurangasiriwardena) [inline components/branding-preference-mgt/org.wso2.carbon.identity.branding.preference.mgt/src/main/java/org/wso2/carbon/identity/branding/preference/mgt/internal/BrandingPreferenceManagerComponent.java:51]: Let's make this a debug log. No need to clutter the startup logs.

File: components/branding-preference-mgt/org.wso2.carbon.identity.branding.preference.mgt/src/main/java/org/wso2/carbon/identity/branding/preference/mgt/util/BrandingPreferenceMgtUtils.java [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+        if (StringUtils.isBlank(stringJSON)) {
+            return false;
+        }
+        try {
+            JSONObject objectJSON = new JSONObject(stringJSON);
+            if (objectJSON.length() == 0) {
+                return false;
+            }
+        } catch (JSONException exception) {
+            return false;
```
Comment (madurangasiriwardena) [inline components/branding-preference-mgt/org.wso2.carbon.identity.branding.preference.mgt/src/main/java/org/wso2/carbon/identity/branding/preference/mgt/util/BrandingPreferenceMgtUtils.java:67]: Would it be better to add a comment and a debug log here?

File: components/branding-preference-mgt/org.wso2.carbon.identity.branding.preference.mgt/src/main/java/org/wso2/carbon/identity/branding/preference/mgt/BrandingPreferenceManagerImpl.java [hunk 3]
```diff
@@ -0,0 +1,324 @@
...
+import static org.wso2.carbon.identity.branding.preference.mgt.util.BrandingPreferenceMgtUtils.handleClientException;
+import static org.wso2.carbon.identity.branding.preference.mgt.util.BrandingPreferenceMgtUtils.handleServerException;
+
+/**
+ * Branding Preference Management service implementation.
+ */
+public class BrandingPreferenceManagerImpl implements BrandingPreferenceManager {
+
+    private static final Log log = LogFactory.getLog(BrandingPreferenceManagerImpl.class);
+    //TODO: Improve API to manage application level & language level theming resources in addition to the tenant level.
```
Comment (madurangasiriwardena) [inline components/branding-preference-mgt/org.wso2.carbon.identity.branding.preference.mgt/src/main/java/org/wso2/carbon/identity/branding/preference/mgt/BrandingPreferenceManagerImpl.java:67]: Let's remove the TODO and create a git issue if this needs to be tracked further.

File: components/branding-preference-mgt/org.wso2.carbon.identity.branding.preference.mgt/src/main/java/org/wso2/carbon/identity/branding/preference/mgt/BrandingPreferenceManagerImpl.java [hunk 4]
```diff
@@ -0,0 +1,324 @@
...
+        try {
+            InputStream inputStream = BrandingPreferenceMgtUtils.generatePreferenceInputStream(preferencesJSON);
+            Resource brandingPreferenceResource = buildResourceFromBrandingPreference(resourceName, inputStream);
+            getConfigurationManager().addResource(BRANDING_RESOURCE_TYPE, brandingPreferenceResource);
+        } catch (ConfigurationManagementException e) {
+            if (RESOURCE_ALREADY_EXISTS_ERROR_CODE.equals(e.getErrorCode())) {
+                if (log.isDebugEnabled()) {
+                    log.debug("Branding preferences are already exists for tenant: " + tenantDomain, e);
+                }
+                throw handleClientException(ERROR_CODE_BRANDING_PREFERENCE_NOT_EXISTS, tenantDomain);
```
Comment (madurangasiriwardena) [inline components/branding-preference-mgt/org.wso2.carbon.identity.branding.preference.mgt/src/main/java/org/wso2/carbon/identity/branding/preference/mgt/BrandingPreferenceManagerImpl.java:94]: Is this error code correct? Error checked in the if condition does not match here.

File: components/branding-preference-mgt/org.wso2.carbon.identity.branding.preference.mgt/src/main/java/org/wso2/carbon/identity/branding/preference/mgt/BrandingPreferenceManagerImpl.java [hunk 5]
```diff
@@ -0,0 +1,324 @@
...
+        }
+
+        try {
+            InputStream inputStream = BrandingPreferenceMgtUtils.generatePreferenceInputStream(preferencesJSON);
+            Resource brandingPreferenceResource = buildResourceFromBrandingPreference(resourceName, inputStream);
+            getConfigurationManager().replaceResource(BRANDING_RESOURCE_TYPE, brandingPreferenceResource);
+        } catch (ConfigurationManagementException e) {
+            throw handleServerException(ERROR_CODE_ERROR_UPDATING_BRANDING_PREFERENCE, tenantDomain, e);
+        } catch (UnsupportedEncodingException e) {
+            throw handleServerException(ERROR_CODE_UNSUPPORTED_ENCODING_EXCEPTION, e);
```
Comment (madurangasiriwardena) [inline components/branding-preference-mgt/org.wso2.carbon.identity.branding.preference.mgt/src/main/java/org/wso2/carbon/identity/branding/preference/mgt/BrandingPreferenceManagerImpl.java:168]: Shouldn't this be a client exception?

**Comment by madurangasiriwardena** [comment]
Please check why the build is failing.

File: components/branding-preference-mgt/org.wso2.carbon.identity.branding.preference.mgt/src/main/java/org/wso2/carbon/identity/branding/preference/mgt/internal/BrandingPreferenceManagerComponent.java (hunk 1 above)
Comment (UdeshAthukorala) [reply components/branding-preference-mgt/org.wso2.carbon.identity.branding.preference.mgt/src/main/java/org/wso2/carbon/identity/branding/preference/mgt/internal/BrandingPreferenceManagerComponent.java]: Addressed with https://github.com/wso2/carbon-identity-framework/pull/3870/commits/d8b98f955a8d30d3fb07768ec4b1e5503aa46a48

File: components/branding-preference-mgt/org.wso2.carbon.identity.branding.preference.mgt/src/main/java/org/wso2/carbon/identity/branding/preference/mgt/util/BrandingPreferenceMgtUtils.java (hunk 2 above)
Comment (UdeshAthukorala) [reply components/branding-preference-mgt/org.wso2.carbon.identity.branding.preference.mgt/src/main/java/org/wso2/carbon/identity/branding/preference/mgt/util/BrandingPreferenceMgtUtils.java:74]: Addressed with https://github.com/wso2/carbon-identity-framework/pull/3870/commits/d8b98f955a8d30d3fb07768ec4b1e5503aa46a48

File: components/branding-preference-mgt/org.wso2.carbon.identity.branding.preference.mgt/src/main/java/org/wso2/carbon/identity/branding/preference/mgt/BrandingPreferenceManagerImpl.java (hunk 4 above)
Comment (UdeshAthukorala) [reply components/branding-preference-mgt/org.wso2.carbon.identity.branding.preference.mgt/src/main/java/org/wso2/carbon/identity/branding/preference/mgt/BrandingPreferenceManagerImpl.java]: Addressed with https://github.com/wso2/carbon-identity-framework/pull/3870/commits/d8b98f955a8d30d3fb07768ec4b1e5503aa46a48

File: components/branding-preference-mgt/org.wso2.carbon.identity.branding.preference.mgt/src/main/java/org/wso2/carbon/identity/branding/preference/mgt/BrandingPreferenceManagerImpl.java (hunk 5 above)
Comment (UdeshAthukorala) [reply components/branding-preference-mgt/org.wso2.carbon.identity.branding.preference.mgt/src/main/java/org/wso2/carbon/identity/branding/preference/mgt/BrandingPreferenceManagerImpl.java]: Addressed with https://github.com/wso2/carbon-identity-framework/pull/3870/commits/d8b98f955a8d30d3fb07768ec4b1e5503aa46a48

**Pull request closed without merging**